### PR TITLE
Added the ability to save keybindings.

### PR DESCRIPTION
**Pull request by lilSander**
I added the ability to save the keybindings in a json file.
I testet it both with the editor and as a standalone program.
But to be safe, it should be double checked, if it also works on other machines.

### CODE REVIEW

File: Assets/SEE/Controls/KeyBindings.cs [hunk 1]
```diff
@@ -311,8 +312,29 @@ public static void SetBindingForKey(KeyActionDescriptor descriptor, KeyCode keyC
...
+            }
+        }
+
+        /// <summary>
+        /// Returns a list of all keyCodes.
+        /// </summary>
+        /// <returns>all keyCodes</returns>
+        internal static List<KeyCode> GetKeyCodes()
+        {
+            List<KeyCode> keys = new List<KeyCode>();
```
Comment (koschke) [inline Assets/SEE/Controls/KeyBindings.cs:332]: You can use the simplified new operator here: `List<KeyCode> keys = new();`

File: Assets/SEE/Controls/KeyBindings.cs [hunk 2]
```diff
@@ -311,8 +312,29 @@ public static void SetBindingForKey(KeyActionDescriptor descriptor, KeyCode keyC
...
+                jsonObject[index] = keyCode;
+                File.WriteAllText(path, JsonConvert.SerializeObject(jsonObject));
+            }
+        }
+
+        /// <summary>
+        /// Returns a list of all keyCodes.
+        /// </summary>
+        /// <returns>all keyCodes</returns>
+        internal static List<KeyCode> GetKeyCodes()
```
Comment (koschke) [inline Assets/SEE/Controls/KeyBindings.cs:330]: Whenever possible, interfaces should be used. Better use `IList` than `List`. If the order does not matter, you could even use `ICollection`.

File: Assets/SEE/Controls/KeyBindings.cs [hunk 3]
```diff
@@ -311,8 +312,29 @@ public static void SetBindingForKey(KeyActionDescriptor descriptor, KeyCode keyC
...
+        {
+            List<KeyCode> keys = new List<KeyCode>();
+            foreach(var binding in keyBindings)
```
Comment (koschke) [inline Assets/SEE/Controls/KeyBindings.cs:333]: I believe this loop can be replaced by `keys.AddRange (...)`.

File: Assets/SEE/UI/SettingsMenu.cs [hunk 4]
```diff
@@ -49,6 +51,31 @@ public class SettingsMenu : PlatformDependentComponent
         /// </summary>
         protected override void StartDesktop()
         {
+            string path = Application.dataPath + "/StreamingAssets/KeyBindings.json";
```
Comment (koschke) [inline Assets/SEE/UI/SettingsMenu.cs:54]: The same `Application.dataPath + "/StreamingAssets/KeyBindings.json"` is used above. This redundancy must be avoided. Use a function instead at both occurrences.

File: Assets/SEE/UI/SettingsMenu.cs [hunk 5]
```diff
@@ -49,6 +51,31 @@ public class SettingsMenu : PlatformDependentComponent
...
+            string allKeys = JsonConvert.SerializeObject(KeyBindings.GetKeyCodes());
+            // when the file exists, read the file, when it does not exist, write to the file, to make reading possible.
+            if (File.Exists(path) && File.ReadAllText(path) != null)
+            {
+                string jsonContent = File.ReadAllText(path);
+                List<KeyCode> keyBindingDataList = JsonConvert.DeserializeObject<List<KeyCode>>(jsonContent);
+                int index = 0;
+                // check if new keyBindings got added, which are not present in the file.
+                if(keyBindingDataList.Count() == KeyBindings.GetKeyCodes().Count())
```
Comment (koschke) [inline Assets/SEE/UI/SettingsMenu.cs:63]: There must be a blank between `if` and `(`.

File: Assets/SEE/UI/SettingsMenu.cs [hunk 6]
```diff
@@ -49,6 +51,31 @@ public class SettingsMenu : PlatformDependentComponent
...
+                {
+                    foreach (var binding in KeyBindings.AllBindings())
+                    {
+                        binding.Value.KeyCode = keyBindingDataList[index++];
+                    }
+                }
+                else
+                {
+                    File.WriteAllText(Application.dataPath + "/StreamingAssets/KeyBindings.json", allKeys);
```
Comment (koschke) [inline Assets/SEE/UI/SettingsMenu.cs:72]: See above. Use a function.

File: Assets/SEE/UI/SettingsMenu.cs [hunk 7]
```diff
@@ -49,6 +51,31 @@ public class SettingsMenu : PlatformDependentComponent
...
+                }
+            }
+            else
+            {
+                File.WriteAllText(Application.dataPath + "/StreamingAssets/KeyBindings.json", allKeys);
```
Comment (koschke) [inline Assets/SEE/UI/SettingsMenu.cs:77]: See above. Use a function.

File: Assets/SEE/UI/SettingsMenu.cs [hunk 8]
```diff
@@ -49,6 +51,31 @@ public class SettingsMenu : PlatformDependentComponent
         /// </summary>
         protected override void StartDesktop()
         {
+            string path = Application.dataPath + "/StreamingAssets/KeyBindings.json";
+            string allKeys = JsonConvert.SerializeObject(KeyBindings.GetKeyCodes());
+            // when the file exists, read the file, when it does not exist, write to the file, to make reading possible.
+            if (File.Exists(path) && File.ReadAllText(path) != null)
+            {
+                string jsonContent = File.ReadAllText(path);
```
Comment (koschke) [inline Assets/SEE/UI/SettingsMenu.cs:59]: Why do you read the file twice? In line 57 and here again. That could be avoided.

File: Assets/SEE/UI/SettingsMenu.cs (hunk 4 above)
Comment (koschke) [inline Assets/SEE/UI/SettingsMenu.cs:54]: Why are lines 54-78 not contained in a method of their own? And this method should belong to class `KeyBindings`. That class should have a load and save method. How the key bindings are stored must be a secret of this class.

File: Assets/SEE/Controls/KeyBindings.cs [hunk 9]
```diff
@@ -311,8 +312,29 @@ public static void SetBindingForKey(KeyActionDescriptor descriptor, KeyCode keyC
             }
             else
             {
+                KeyCode oldKey = descriptor.KeyCode;
                 descriptor.KeyCode = keyCode;
+                string path = Application.dataPath + "/StreamingAssets/KeyBindings.json";
```
Comment (koschke) [inline Assets/SEE/Controls/KeyBindings.cs:317]: This code must become a method of class `KeyBindings`. See my comment below.

Moreover, the key binding is a mapping of keys onto actions. You are just writing the keys and assume their number stays the same and that there is always a fixed order. Neither should be asumed. What do you do if we later add or remove a key binding? Furthermore, the file in which the key bindings are stored must be understood and possibly edited by a human. Hence, the interpretation of its content must be simple and clear. The content should be a JSON structure of key-value pairs. The JSON key is the key code and the JSON value is the action.

**Review by koschke — CHANGES_REQUESTED** [review]
See my detailed comments. The main issue is that saving/storing must be methods of `KeyBindings` and the syntax of the JSON file must be understood by a human and be tolerant against changes (new or removed key bindings).

File: Assets/SEE/Controls/KeyBindings.cs (hunk 9 above)
Comment (lilSander) [reply Assets/SEE/Controls/KeyBindings.cs:317]: Well, the case that key bindings get removed or added, is handelt by keyBindingDataList.Count() == KeyBindings.GetKeyCodes().Count(), but yes if someone changes the order of the bindings, then I have a big problem. Im going to fix it.

**Comment by falko17** [comment]
The pipeline failed because our CI servers `see-ci1` and `see-ci2` got rebooted at some point, which meant the GitHub runner wasn't active. I have installed the runner as a service on both machines now, so that the runner starts automatically when the server is started.

**Comment by falko17** [comment]
@koschke Now it seems that dashboard tests are failing because our Dashboard instance at stvr2 is down. Maybe we should allow failure for these tests (in the CI at least), as they are very reliant on external conditions?

**Comment by koschke** [comment]
The server for the Axivion Dashboard is up and running again. @falko17 Yes, we could disable those tests.

File: Assets/SEE/Controls/KeyBindings.cs (hunk 9 above)
Comment (lilSander) [reply Assets/SEE/Controls/KeyBindings.cs]: @koschke please check again, if its better now.

Comment (koschke) [reply Assets/SEE/Controls/KeyBindings.cs]: Much better now.

**Comment by koschke** [comment]
I update the public key of the Axivion server in the DashboardRetriever. Tests are running again.

**Review by koschke — APPROVED** [review]
I refactored the code quite a bit; see my commit messages for details.
I added tests.

Can be merged now.

**Pull request merged**